rfcs/2020-05-25-2685-dev-workflow-simplification.md

```diff
@@ -0,0 +1,244 @@
+
+If you're interested in having Nix **automatically start your environment**, you can consider a tool like [`direnv`](https://direnv.net/docs/installation.html), often available from your package manager.
+
+```bash
+❯ cd /git/timberio/vector
+❯ direnv allow .
+❯ cd ../vector
+direnv: loading /git/timberio/vector/.envrc
+direnv: using nix
+cdirenv: export +AR +AS +CC +CONFIG_SHELL +CXX +HOST_PATH +IN_NIX_SHELL +LC_ALL +LD +NIX_BINTOOLS +NIX_BINTOOLS_WRAPPER_x86_64_unknown_linux_gnu_TARGET_HOST +NIX_BUILD_CORES +NIX_BUILD_TOP +NIX_CC +NIX_CC_WRAPPER_x86_64_unknown_linux_gnu_TARGET_HOST +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_STORE +NM +OBJCOPY +OBJDUMP +PROTOC +PROTOC_INCLUDE +RANLIB +READELF +SIZE +SOURCE_DATE_EPOCH +STRINGS +STRIP +TEMP +TEMPDIR +TMP +TMPDIR +buildInputs +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +name +nativeBuildInputs +nobuildPhase +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~LOCALE_ARCHIVE ~PATH
+```
+
+Now you can use the jobs detailed in **"Bring your own toolbox"** below. 
+
+### Bring your own toolbox
+
+> **Targets:** This option is required for MSVC/Mac/FreeBSD toolchains.
+
+To build Vector on your own host will require a fairly complete development environment!
+
+We keep an up to date list of all dependencies used in our CI environment inside our `default.nix` file. Loosely, you'll need the following:
+
+- **To build Vector:** Have working Rustup, C++/C build tools (LLVM, GCC, or MSVC), Python, and Perl, `make` (the GNU one preferably), `bash`, `cmake`, and `autotools`
+- **To run integration tests:** Have `docker` available, or a real live version of that service.
+- **To build the Website:** Have a working modern Ruby and Bundler toolchain available, also `bundle install` in the `scripts/` directory.
+- **To run the Website in Dev:** Have a working `node` environment with `npm`/`yarn`, also run `yarn` from the `website/` directory.
+- **To run `make check-component-features`:** Have `remarshal` installed.
+
+If you find yourself needing to run something (such as `make generate`) inside the Docker environment described above, that's totally fine, they won't collide or hurt each other. In this case, you'd just run `make environment-generate`.
+
+We're interested in reducing our dependencies if simple options exist. Got an idea? Try it out, we'd to hear of your successes and failures!
+
+In order to do your development on Vector, you'll primarily use a few commands, ordered from most to least frequently run:
+
+```bash
+# Validate your code can compile
+make check
```

Comment:
Do we still want to use `make` in these situations? Especially when you've built your own dev env, it'd be nice to get rid of this extra layer if it's not adding anything over `cargo`.

Reply:
I think we don't need targets that are direct counterparts of the *simple* cargo commands, like `check`, `build` and `test`.

But then, if we omit them - `make` is no longer "the single entry point". Frankly, I'm not a big fan of `make` being used like this either. I prefer executable like `bin/test-integration`. With that approach, there's not so much contradiction, cause it feels more natural to use `cargo test` alongside `bin/test-integration ...`.

Reply:
Yeah I had basically the same conflict. Should we get rid of these jobs or just document that cargo works fine too?

Reply:
Keeping them makes sense from the  "single entry" point of view. I wonder how it would work if we remove them though. I won't miss them. Let's try removing them and see what it does?

Reply:
I'd miss them. :)